### PR TITLE
feat: add test operation

### PR DESCRIPTION
**Pull request by steffenix**
@dougstorm @fubuloubu @fameal  your feedback would be appreciated on this draft.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -1,117 +1,57 @@
...
-    bal = token.totalSupply() // 10000
-    token.transfer(a, bal, {"from": andre})
-    # Unlimited Approvals
-    token.approve(vault, 2 ** 256 - 1, {"from": a})
-    # Deposit half their stack
-    vault.deposit(bal // 2, {"from": a})
-    yield a
+def token():
+    token_address = "0x6b175474e89094c44da98b954eedeac495271d0f"  # this should be the address of the ERC-20 used by the strategy/vault
+    yield Contract.from_explorer(token_address)
```
Comment (Macarse) [inline tests/conftest.py:39]: `Contract(token_address)` since we have `autofetch_sources: True`

File: tests/conftest.py [hunk 2]
```diff
@@ -1,117 +1,57 @@
...
-    token.transfer(a, bal, {"from": andre})
-    # Unlimited Approvals
-    token.approve(vault, 2 ** 256 - 1, {"from": a})
-    # Deposit half their stack
-    vault.deposit(bal // 2, {"from": a})
-    yield a
+def strategy(strategist, keeper, vault, Strategy, gov):
+    strategy = strategist.deploy(Strategy, vault)
+    strategy.setKeeper(keeper)
+    vault.addStrategy(strategy, 10_000, 10_000, 10, {"from": gov})
```
Comment (Macarse) [inline tests/conftest.py:56]: I would prefer:
`vault.addStrategy(strategy, 10_000, 0, 1_000, {"from": gov})`

File: tests/test_operation.py [hunk 3]
```diff
@@ -1,8 +1,25 @@
...
-
-def test_operation():
-    pass
+from brownie import Contract
+
+
+def test_operation(accounts, token, vault, strategy, strategist):
+    # First you need to get some funds for the token you are about to use,
+    # in this example it impersonate an exchange address to use it's funds.
+    amount = 10_000 * 10e18
```
Comment (Macarse) [inline tests/test_operation.py:7]: `token.decimals()`?

**Review by Macarse — APPROVED** [review]

File: tests/conftest.py (hunk 1 above)
Comment (fubuloubu) [reply tests/conftest.py:39]: This change will make sure that contracts pulled from Etherscan are cached
```suggestion
    yield Contract(token_address)
```

File: tests/conftest.py (hunk 2 above)
Comment (steffenix) [reply tests/conftest.py:56]: ok, could you explain why?

Comment (Macarse) [reply tests/conftest.py]: rate = 0 is unlimited.

**Comment by steffenix** [comment]
Things I am not sure about?
- Should I assert anything after calking `tend`?
- Is there a better test than checking balance isn't zero after withdrawal?

I will take time tomorrow to understand and add test for the emergency.
Any other tests that could be added?

**Comment by fameal** [comment]
#14

**Comment by fubuloubu** [comment]
@steffenix can you copy/adapt the `test.yml` github actions from `vaults`? (I think some changes are necessary)

**Comment by steffenix** [comment]
Yes will do, but will need someone from your team that setup the env variables.
- ETHERSCAN_TOKEN
- WEB3_INFURA_PROJECT_ID

**Comment by steffenix** [comment]
@fubuloubu added test and linting copied from https://github.com/iearn-finance/yearn-protocol/tree/develop/.github/workflows

**Comment by fameal** [comment]
Why don't you do the same thing Carlos did here: https://github.com/iearn-finance/yearn-vaults/pull/35 to get 2 runs instead of 4

**Comment by steffenix** [comment]
> Why don't you do the same thing Carlos did here: [iearn-finance/yearn-vaults#35](https://github.com/iearn-finance/yearn-vaults/pull/35) to get 2 runs instead of 4

The PR you pointed me to changes the branches CI runs on, not the number of jobs.
Plus I would say it's good to have multiple jobs, it runs faster.

**Comment by fubuloubu** [comment]
> > Why don't you do the same thing Carlos did here: [iearn-finance/yearn-vaults#35](https://github.com/iearn-finance/yearn-vaults/pull/35) to get 2 runs instead of 4
> 
> The PR you pointed me to changes the branches CI runs on, not the number of jobs.
> Plus I would say it's good to have multiple jobs, it runs faster.

The only trick is that code coverage doesn't work across multiple jobs, but I don't think coverage is necessary for this repo anyways.

**Comment by steffenix** [comment]
> > > Why don't you do the same thing Carlos did here: [iearn-finance/yearn-vaults#35](https://github.com/iearn-finance/yearn-vaults/pull/35) to get 2 runs instead of 4
> > 
> > 
> > The PR you pointed me to changes the branches CI runs on, not the number of jobs.
> > Plus I would say it's good to have multiple jobs, it runs faster.
> 
> The only trick is that code coverage doesn't work across multiple jobs, but I don't think coverage is necessary for this repo anyways.

I didn't enabled test coverage on this repo because that can be up to the people that will use the strategy to build relevant tests and decide if they want to use coverage.

**Review by fubuloubu — APPROVED** [review]

**Comment by fubuloubu** [comment]
oh @steffenix I should have had you add to the README about the environment variables, so people can run tests in their forks

**Pull request merged**